sksports/model/heart_rate.py

```diff
@@ -0,0 +1,164 @@
+"""Module to estimate heart rate from data."""
+
+# Authors: Aart Goossens
+#          Guillaume Lemaitre
+# License: MIT
+
+from __future__ import division
+
+import numpy as np
+from scipy.optimize import leastsq, minimize
+
+from sklearn.base import BaseEstimator, RegressorMixin
+from sklearn.utils import check_X_y, check_array
+from sklearn.utils.validation import check_is_fitted
+
+
+def _model(power, hr_rest, hr_max, hr_slope, hr_drift, rate_growth,
+           rate_decay):
+    """Model to predict heart-rate from power data."""
+    power_corrected = power + power.cumsum() * hr_drift
+    hr_lin = power_corrected * hr_slope + hr_rest
+    hr_ss = np.minimum(hr_max, hr_lin)
+
+    diff_hr = np.pad(np.diff(power_corrected), pad_width=1, mode='constant')
+    rate = np.where((diff_hr <= 0), rate_decay, rate_growth)
+
+    hr_pred = np.ones(power.shape) * hr_rest
+    for curr_idx in enumerate(hr_pred.size):
+        hr_prev = hr_pred[curr_idx] if curr_idx > 0 else hr_rest
+        hr_pred[curr_idx] = (hr_prev +
+                             (hr_ss[curr_idx] - hr_prev) / rate[curr_idx])
+
+    return hr_pred
+
+
+def _model_residual(params, heart_rate, power):
+    """Residuals between prediction and ground-truth."""
+    return heart_rate - _model(power, *params)
+
+
+def _model_residual_least_square(params, heart_rate, power):
+    """Least-square error of the residuals."""
+    return np.sum(_model_residual(params, heart_rate, power) ** 2)
+
+
+class HeartRateRegressor(BaseEstimator, RegressorMixin):
```

Comment:
Can you explain the benefit of `BaseEstimator` and `RegressorMixin`? I don't see it used anywhere in the `fit` and `predict` methods. Is there other functionality of these classes that can be useful?
Either way, I like implementing this model like a regressor class, it makes the api much more clear.

Reply:
basically, it allows to use the tool as pipeline, grid-search, and cross-validation from scikit-learn.
The mixin just define a `score` function which is the R^2 score while the `BaseEstimator` provides `get_params` and `set_params` to deal with the parameters.

Reply:
> Either way, I like implementing this model like a regressor class, it makes the api much more clear.

We can also the model public to the user. We just need a good name.